### ISSUE TITLE
Provide a way to opt out of CUDA lock arrays without RDC

### ISSUE THREAD

**Issue by ibaned** [kokkos/kokkos#1375]
Due to #1350, some applications including a few of my own would prefer not to have to use RDC for best performance, since that would mean losing all the benefits of shared libraries. Having a way to simply opt out of the features supported by lock arrays would be ideal in this case. One idea originally proposed by @dsunder would be template parameters along the lines of:

```cpp
Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::NativeSizedAtomicsOnly, Kokkos::NoLevel1Scratch>
```

I would actually be in favor of something that is a little bit more implementation-specific but honest about its purpose:

```cpp
Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::NoLockArrays>
```

We've considered a compile-time switch that disables them for all parallel regions, but some important kernels in Trilinos do use these features and so its better to be able to have a few parallel regions still use them.

**Comment by mhoemmen** [comment]
What effect would `NoLevel1Scratch` have?  Would that affect performance of RangePolicy reductions?

**Comment by ibaned** [comment]
There is a second lock array (besides the one used to implement atomics on types bigger than 64 bits) which is used to throttle the amount of parallelism on the GPU in order to avoid having to allocate too much level 1 scratch memory at once. Our initial thinking was that this template option would disallow level 1 scratch use altogether, I'm not sure if there are scenarios in which we could just turn of the throttling and allow certain amounts of level 1 scratch. As far as I know this mechanism isn't used by reductions.

**Comment by ibaned** [comment]
Actually, to be more clear, that second array only affects `parallel_for` with the `TeamPolicy` (not the simple `RangePolicy`), in the `Kokkos::Cuda` space, with level 1 scratch. Thats possibly another argument for the `Kokkos::NoLockArrays` version, because any other use of level 1 scratch seems to be fine

**Comment by crtrott** [comment]
I understand the desire for this, but I am a bit hesitant to going down the road ("bit hesitant" means their is a fair chance you can convince me with enough whining about poor performance ;-) ). My main reason to be hesitant is that I don't want to start adding all those options for turning features on and off. Keeping track of all the possible combinations and making sure that it all works is gonna be difficult. Furthermore we would be adding options which allow people to shoot themselves in the foot in particular in larger development teams. In a lot of our large applications we are calling through deep stacks of functionals within the kernel. How does someone on the inner layer notice that this option NoLockArray is actually set for the policy. Sure he could try and track first all the places down from which his function might be called inside a parallel for, and then track down the typedef for the policy being used, but even so does that mean he understands that NoLockArrays means he can't use an exchange atomic on a complex<double>? Furthermore setting this option will actually still work in many many cases, because most of the time the symbols will still contain the right value. But once in a while they will not, which may depend on the hardware people are run on and the GPU driver version for example. Debugging this would be a pain. And other than a global option there is no way we can detect improper use at compile or even runtime.

**Comment by crtrott** [comment]
I am remarking this as a question until we have a decision on whether we go down this road.

**Comment by mhoemmen** [comment]
Part of the issue is Trilinos' current `CUDA_LAUNCH_BLOCKING=1` requirement.  If kernel launches and memcpys were asynchronous by default, perhaps Trilinos users wouldn't see the cost of copying 24 bytes to the GPU quite so much.   Removing this requirement would take a lot of effort and testing fully downstream (including in applications), though, and I don't see any plans this fiscal year for doing so.

**Comment by crtrott** [comment]
Mark that wouldn't change a thing on that front. That particular copy does not exist as an asynchronous version ...

**Comment by ibaned** [comment]
@mhoemmen also we've got data from applications (BoxLib) that did not require UVM or `CUDA_LAUNCH_BLOCKING`, and still saw a 50% slowdown from having lock arrays copied.

Given that the advantage of GPUs over CPUs still hovers at around 2X, a 1.5X slowdown is not really acceptable at all. @crtrott and I did some experiments with miniFE that showed overheads of around 30% for small problems. Although these overheads did amortize to insignificant percentages for bigger problems, I do know of applications where there are many smaller kernels that will still feel large overheads. Personally, all I care about is getting rid of this overhead. I'd love to use RDC but apparently NVIDIA didn't bother implementing their shared linker correctly, so that forces people to take on the compilation headaches of static libraries (Albany+Trilinos take much longer to compile that way). It seems very easy to disable these lock arrays one way or another inside Kokkos, so I think its a worthwhile thing to discuss and implement.

The risk that I see with not doing this is that Kokkos exists purely for performance reasons, and if some part of it is slower than it needs to be and can be easily hacked around, users will start doing that to get the performance results they need (or they'll switch to another tool). We've seen users do things like get raw pointers and use those if View indexing is a bit too slow, and call `Experimental::md_parallel_for` just because it doesn't update the lock arrays correctly.

I do agree with @crtrott that doing this on a per-policy basis introduces the risk of mistakes that are hard to debug, and that a global compile macro would at least be easier to safety check. However, the global option will be useless if these features do get used in some critical piece of the code.

I personally think that large atomics are rarely used besides complex, and complex atomic addition should be implemented on a per-component basis.

We can actually pretty safely check the level 1 scratch thing on a per-kernel basis (if policy has trait and level 1 scratch requested, error out), the only thing thats hard to check is the large atomics. One possible plan of action is this:

0. Make sure NVIDIA is aware their shared linker is broken
1. Implement complex atomic add and assign as per-component operations
2. Allow global macro to disable all other large atomics
3. check if Trilinos with complex enabled still works in that configuration (maybe @mhoemmen knows if there is a reason it won't)
4. introduce a policy trait to disable L1 scratch

**Comment by mhoemmen** [comment]
@ibaned wrote:
> check if Trilinos with complex enabled still works in that configuration (maybe @mhoemmen knows if there is a reason it won't)

Sacado and Stokhos require large atomics.  They instantiate Tpetra and solver objects.

There may be the occasional kernel that does `Kokkos::pair` atomics.

It should be straightforward for Tpetra and downstream solvers to decide, per kernel, at compile time, whether that kernel needs large atomics.  Is that OK?

**Comment by ibaned** [comment]
> It should be straightforward for Tpetra and downstream solvers to decide, per kernel, at compile time, whether that kernel needs large atomics. Is that OK?

Feasibility, yes. However @crtrott 's concern is that it will be hard to enforce. Meaning if you mark your kernel with `Kokkos::NoLargeAtomics` and you accidentally use a large atomic, it may work sometimes and fail nondeterministically. Actually, now that I think about this, one thing we could do is that if `KOKKOS_DEBUG` is on then we still do the `cudaMemcpyToSymbol`, but this time we force the values to be `nullptr`. That will at least ensure any and all attempts fail, giving users a path forward to debug this.

**Comment by mhoemmen** [comment]
@ibaned wrote:
> ... if you mark your kernel with Kokkos::NoLargeAtomics and you accidentally use a large atomic, it may work sometimes and fail nondeterministically.

I'm OK with this.  It's an optimization, and you have to opt into it.

This is one of the reasons I wish Kokkos' atomics were tied to the parallel loop policy.  If they were, we could simply forbid unsupported types at compile time.  Developers would get a `static_assert` with a clear error message ("Your parallel loop does not support atomics with general types; please change the range or team policy accordingly").

**Comment by mhoemmen** [comment]
I mean, why _aren't_ Kokkos' atomics tied to the execution space at least?  It seems like that would make it much easier to implement them for new execution spaces.

**Comment by ibaned** [comment]
@mhoemmen I agree. In retrospect, having the execution space or even the whole policy be a required template argument to the atomic functions would have definitely simplified their implementation. At this point it would be backwards-compatibility-breaking, but it may still make sense to introduce that as a new option and get people to start using it.

Until that exists and people are mostly using it, we have to deal with the strange and implicit nature of atomics. That said, the method I described above of writing in a null pointer when `KOKKOS_DEBUG` is `ON` might address @crtrott 's concerns regarding debuggability.

**Comment by mhoemmen** [comment]
> In retrospect, having the execution space or even the whole policy be a required template argument to the atomic functions would have definitely simplified their implementation.

It would also have allowed for architectures on which inter-team atomics don't exist or are slow, while intra-team atomics exist or are fast.

**Comment by crtrott** [comment]
Regarding making the policy a required arg for the atomics: i bet Si would live that, since now all functions which may use atomics or call functions which may use atomics must be templated on the policy type 😳

**Comment by mhoemmen** [comment]
@crtrott wrote:
> i bet Si would love that, since now all functions which may use atomics or call functions which may use atomics must be templated on the policy type

Ha! ;-P  It's not that bad though.  This is a good reason to have a "union execution space" for all CPU-based execution spaces.

Does Kokkos promise that Kokkos atomics mean anything outside of a Kokkos parallel region?

**Comment by ibaned** [comment]
@crtrott good point, for non-trivial kernels that gets ugly. This also includes the `operator()` of any atomic views or the `ScatterView`.

**Comment by crtrott** [comment]
Here is another thought: can you check what happens performance-wise if we store these guys simply in __device__ symbols not in __constant__ symbols? After talking to NVIDIA folks at the meeting I am at right now, it seems like the __device__ symbols have a different scoping than __constant__ symbols. In the end it means we need to update the device symbols only once after entering the object file, not every time we launch a kernel. We could simply have a file scope static host side variable which serves as a flag whether we updated this thing ever.

**Comment by crtrott** [comment]
I am closing this, addressed this in a better way in #1385
